test_util/helpers.py

```diff
@@ -23,81 +24,137 @@ def path_join(p1, p2):
+    """URL abstraction to allow convenient substitution of URL anatomy
+    """
+    def __init__(self, scheme, host, path, query, fragment, port):
+        self.scheme = scheme
+        self.host = host
+        self.path = path
+        self.query = query
+        self.fragment = fragment
+        self.port = port
+
+    @classmethod
+    def from_string(cls, url_str):
+        u = urlsplit(url_str)
+        if ':' in u.netloc:
+            host, port = u.netloc.split(':')
+        else:
+            host = u.netloc
+            port = None
+        return cls(u.scheme, host, u.path, u.query, u.fragment, port)
+
+    @property
+    def netloc(self):
+        return '{}:{}'.format(self.host, self.port) if self.port else self.host
+
+    def __str__(self):
+        return urlunsplit((
+            self.scheme,
+            self.netloc,
+            self.path,
+            self.query if self.query else '',
+            self.fragment if self.fragment else ''))
+
+    def get_url(self, scheme=None, host=None, path=None, query=None, fragment=None, port=None):
+        """return new Url with any component replaced
+        """
+        return Url(
+            scheme if scheme is not None else self.scheme,
+            host if host is not None else self.host,
+            path if path is not None else self.path,
+            query if query is not None else self.query,
+            fragment if fragment is not None else self.fragment,
+            port if port is not None else self.port)
+
 
-    def __init__(self, default_host_url, api_base, default_headers=None,
-                 ca_cert_path=None, get_node_url=None):
-        self.default_host_url = default_host_url
-        self.api_base = api_base
-        if default_headers is None:
-            default_headers = dict()
-        self.default_headers = default_headers
-        self.ca_cert_path = ca_cert_path
-        self._get_node_url = get_node_url
-
-    def api_request(self, method, path, host_url=None, node=None, port=None, **kwargs):
+class ApiClient:
+    """utilizes requests.session with some URL handling to remove boilerplate for making
+    cluster requests. Can also be used to bind advanced helper methods of a given service
+    """
+    def __init__(self, url: Url, get_node_port=None):
         """
-        Makes a request with default headers + user auth headers if necessary
-        If self.ca_cert_path is set, this method will pass it to requests
         Args:
-            method: (str) name of method for requests.request
-            path: see get_url()
-            host_url: override the client's default host URL
-            node: if a get_node_url function is set, node and port will be passed to this function
-                instead of setting a host_url or using the default
-            port: can be used with node with get_node_url is set
-            **kwargs: any optional arguments to be passed to requests.request
+            url: Url object to which requests will be made
```

Comment:
So by looking at the code below here it seems to me that the ApiClient will _always_ use this url as full URL, optionally extended with a `path_extension`, and optionally overridden with a different scheme or query etc.

So, it's not really the "Url object to which requests will be made", but more something in the lines of "default URL used by this client, optionally prepended with a path extension or modified upon `client.api_request()` invocation."

Can we call it `default_url`? :)